htlcswitch/failure.go

```diff
@@ -66,6 +70,16 @@ func (l *LinkError) WireMessage() lnwire.FailureMessage {
 	return l.msg
 }
 
+// SourceIndex returns the index of the node that was the source of
+// the error. If this value is 0, the error originated at our node.
+//
+// Note this is part of the ClearTextError interface.
+func (l *LinkError) SourceIndex() int {
+	// LinkErrors originate at our node, so will always have a zero source
+	// index.
+	return 0
```

Comment:
From our perspective isn't the index 0 because we failed the htlc? But I see your argument that this principle only really applies to our own payments, so happy to remove this if it is clearer. I liked this because it makes the router ignorant of the types of errors in the switch (beyond the concept of `ClearTextError`) but this was just a nit on the last PR that I thought I might as well follow up with here.

Reply:
It is a matter of definition. But so far we've only defined the failure source index as an index in the original, full route. It is a value that none of the nodes except the origin node can know.

Reply:
Ok that makes sense. I'll remove this rather than expand the meaning of the value unnecessarily 👍

Reply:
Would it be possible to return an `ErrUnknownSourceIndex` in the cases where we are part of a route but don't know the error source index?

Reply:
The source index is really only applicable in the sender context. Ideally we'd use a data structure for non-sender nodes that doesn't contain this field at all, to prevent any misuse.

Reply:
Yeah, could return an error in that case, to avoid non-senders using it.

Reply:
Going to remove this commit entirely, it makes more sense to restrict the definition of `SourceIndex` to its original meaning.